addons/data-category-tweaks/userscript.js

```diff
@@ -0,0 +1,178 @@
+          // "Make a Variable" always goes first.
+          before.push(blockXML);
+        } else {
+          // "set variable to", etc. always go after.
+          after.push(blockXML);
+        }
+      }
+
+      const result = before;
+
+      if (global.length) {
+        result.push(makeLabel("for-all-sprites"));
+        fixGaps(global);
+        result.push(...global);
+      }
+
+      if (local.length) {
+        result.push(makeLabel("for-this-sprite-only"));
+        fixGaps(local);
+        result.push(...local);
+      }
+
+      return result.concat(after);
+    };
+
+    return separateVariablesByScope(variables).concat(separateVariablesByScope(lists));
+  };
+
+  const injectWorkspace = () => {
+    if (injected) {
+      return;
+    }
+    injected = true;
+
+    const workspace = Blockly.getMainWorkspace();
+    if (!workspace) throw new Error("expected workspace");
+
+    const vm = addon.tab.traps.onceValues.vm;
+    if (!vm) throw new Error("expected vm");
+
+    const flyout = workspace.getFlyout();
+    if (!flyout) throw new Error("expected flyout");
+
+    const DataCategory = workspace.toolboxCategoryCallbacks_.VARIABLE;
+    if (!DataCategory) throw new Error("expected data category");
+
+    let variableCategory;
+    let listCategory;
+
+    const variableCategoryCallback = (workspace) => {
+      let result = DataCategory(workspace);
+
+      if (addon.settings.get("separateLocalVariables")) {
+        result = separateLocalVariables(workspace, result);
+      }
+
+      if (!hasSeparateListCategory) {
+        return result;
+      }
+
+      const { variables, lists } = separateVariablesByType(result);
+      variableCategory = variables;
+      listCategory = lists;
+      return variableCategory;
+    };
+
+    const listCategoryCallback = () => {
+      // Computed in variable category callback
+      return listCategory;
+    };
+
+    const oldShow = flyout.show;
+    flyout.show = function (xml) {
+      this.workspace_.registerToolboxCategoryCallback("VARIABLE", variableCategoryCallback);
+      this.workspace_.registerToolboxCategoryCallback("LIST", listCategoryCallback);
+      oldShow.call(this, xml);
+    };
+
+    // We use Scratch's extension category mechanism to replace the data category with our own.
+    // https://github.com/LLK/scratch-gui/blob/ddd2fa06f2afa140a46ec03be91796ded861e65c/src/containers/blocks.jsx#L344
+    // https://github.com/LLK/scratch-vm/blob/a0c11d6d8664a4f2d55632e70630d09ec6e9ae28/src/engine/runtime.js#L1381
+    const originalGetBlocksXML = vm.runtime.getBlocksXML;
+    vm.runtime.getBlocksXML = function (target) {
+      const result = originalGetBlocksXML.call(this, target);
+      hasSeparateListCategory = addon.settings.get("separateListCategory");
+      if (hasSeparateListCategory) {
+        result.push({
+          id: "data",
+          xml: `
+          <category
+            name="%{BKY_CATEGORY_VARIABLES}"
```

Comment:
Whoops looks like this needs to be a "$"

Reply:
% is correct. Blockly will parse and convert that to a translated string (https://github.com/LLK/scratch-gui/blob/ddd2fa06f2afa140a46ec03be91796ded861e65c/src/lib/make-toolbox-xml.js#L694-L704)

Reply:
Ah... I see